### DISCUSSION TITLE
Mesh_3: should initialize_triangulation_from_labeled_image be a static function of image domains?

### DISCUSSION THREAD

**Discussion by lrineau** [CGAL/cgal#7537] [Ideas]
> @lrineau I am wondering if it would make any sense if the function `initialize_triangulation_from_labeled_image()` was also a static member function of `Labeled_mesh_domain_3`, just like `create_gray_image_mesh_domain()`?

_Originally posted by @afabri in https://github.com/CGAL/cgal/issues/7445#issuecomment-1599217046_

**Comment by lrineau** [comment]
For the moment, the API of that function is:
```c++
  C3t3 c3t3;
  initialize_triangulation_from_labeled_image(c3t3,
                                              domain,
                                              image,
                                              criteria,
                                              static_cast<unsigned char>(0));
```

Your proposal is probably something like:
```c++
C3t3 c3t3 = Domain_type::initialize_triangulation_from_labeled_image(
                                              domain,
                                              image,
                                              criteria,
                                              static_cast<unsigned char>(0));
```

My first remark is that it should not be a **static** function, if the first parameter is the domain itself. No, let's turn that into a non-static member function:
```c++
C3t3 c3t3 = domain.initialize_triangulation_from_labeled_image(
                                              image,
                                              criteria,
                                              static_cast<unsigned char>(0));
```

My next remark is that the `C3t3` is not know by the domain class, and so it should be a non-static member function **template**, used that way:
```c++
C3t3 c3t3 = domain.initialize_triangulation_from_labeled_image<C3t3>(
                                              image,
                                              criteria,
                                              static_cast<unsigned char>(0));
```
and maybe a better name:
```c++
C3t3 c3t3 = domain.initialize_c3t3_from_labeled_image<C3t3>(
                                              image,
                                              criteria,
                                              static_cast<unsigned char>(0));
```

It would be a domain-specific member function template, and an equivalent of the gray-level images domain class.

Anyway, it should not be part of PR #7445, that is already too big to be reviewed efficiently, but in another pull-request.

**Comment by lrineau** [comment]
Actually, `labeled_image` is already part of the domain class name. So we can shorten the name to:

```c++
C3t3 c3t3 = domain.initialize_c3t3<C3t3>(image, criteria, static_cast<unsigned char>(0));
```

@janetournois @afabri

**Reply by afabri** [reply]
Wouldn't it make sense to pass a reference to a `C3t3` object to the function.  We then don't need the `<C3t3>`.
Jane told me that you probably also can get rid of the last parameter.
